numba/typing/arraydecl.py

```diff
@@ -257,19 +253,36 @@ def resolve_nonzero(self, ary, args, kws):
     @bound_function("array.reshape")
     def resolve_reshape(self, ary, args, kws):
         assert not kws
-        shape, = args
-        shape = normalize_shape(shape)
-        if shape is None:
-            return
-        if ary.layout == "C":
-            # Given order='C' (the only supported value), a C-contiguous
-            # array is always returned for a C-contiguous input.
-            layout = "C"
+        if ary.layout not in 'CF':
+            # only work for contiguous array
+            raise TypeError("reshape() supports contiguous array only")
+
+        if len(args) == 1:
+            # single arg
+            shape, = args
+
+            if shape in types.number_domain:
```

Comment:
Hmm, I don't think there's a reason to accept floats and complex numbers, so let's check for types.Integer instead perhaps?